src/core/qgsmaprendererjob.cpp

```diff
@@ -354,6 +377,144 @@ LayerRenderJobs QgsMapRendererJob::prepareJobs( QPainter *painter, QgsLabelingEn
+            continue;
+          }
+          lit->second.push_back( sourceLayerId );
+        }
+      }
+      if ( ! masks->isEmpty() )
+        layerHasMask.insert( sourceLayerId );
+    };
+
+    collectMasks( &labelMasks, labelLayerId );
+    collectMasks( &symbolLayerMasks, vl->id() );
+  }
+
+  if ( maskedSymbolLayers.isEmpty() )
+    return secondPassJobs;
+
+  // Now that we know some layers have a mask, we have to allocate a mask image and painter
+  // for them in the first pass job
+  for ( LayerRenderJob &job : firstPassJobs )
+  {
+    QgsMapLayer *ml = job.layer;
+
+    if ( job.img == nullptr )
+    {
+      job.context.setPainter( allocateImageAndPainter( ml->id(), job.img ) );
+    }
+    if ( layerHasMask.contains( ml->id() ) )
+    {
+      // Note: we only need an alpha channel here, rather than a full RGBA image
+      job.context.setMaskPainter( allocateImageAndPainter( ml->id(), job.maskPass.image ) );
+      job.maskPass.image->fill( 0 );
+    }
+  }
+
+  // Allocate an image for labels
+  if ( labelJob.img == nullptr )
+  {
+    labelJob.img = allocateImage( QStringLiteral( "labels" ) );
+  }
+
+  // Prepare a mask image for labels if needed
+  if ( layerHasMask.contains( labelLayerId ) && labelJob.maskImage == nullptr )
+  {
+    labelJob.context.setMaskPainter( allocateImageAndPainter( QStringLiteral( "label mask" ), labelJob.maskImage ) );
+    labelJob.maskImage->fill( 0 );
+  }
+
+  // Prepare second pass jobs
+  for ( LayerRenderJob &job : firstPassJobs )
+  {
+    QgsMapLayer *ml = job.layer;
+
+    auto it = maskedSymbolLayers.find( ml->id() );
+    if ( it == maskedSymbolLayers.end() )
+      continue;
+
+    const QList<QString> &layerIdList = it->second;
+    const QSet<QgsSymbolLayerId> &symbolList = it->first;
+
+    // copy the initial job ...
+    secondPassJobs.append( LayerRenderJob() );
+    LayerRenderJob &job2 = secondPassJobs.last();
+    job2 = job;
+    job2.cached = false;
+    job2.maskPass.firstPassJob = &job;
+    QgsVectorLayer *vl1 = qobject_cast<QgsVectorLayer *>( job.layer );
+
+    // ... but clear the image
+    job2.context.setMaskPainter( nullptr );
+    job2.context.setPainter( allocateImageAndPainter( vl1->id(), job2.img ) );
+    if ( ! job2.img )
+    {
+      secondPassJobs.removeLast();
+      continue;
+    }
+
+    // Set the list of layer masks for this second pass job.
+    // This will be needed during the second pass composition.
+    for ( QString layerId : layerIdList )
+    {
+      job2.maskPass.maskJobs.push_back( layerId == labelLayerId ? nullptr : layerJobMapping[layerId] );
+    }
+
+    // FIXME: another possibility here, to avoid allocating a new map renderer and reuse the one from
+    // the first pass job, would be to be able to call QgsMapLayerRenderer::render() with a QgsRenderContext.
+    QgsVectorLayerRenderer *mapRenderer = static_cast<QgsVectorLayerRenderer *>( vl1->createMapRenderer( job2.context ) );
```

Comment:
I wonder how this approach works with "non-deterministic" renderers, like the point displacement renderer. (I suspect there will be issues, but it's going to be tricky to avoid that)

Reply:
Ouch ! I did not know there were non-deterministic renderers ... do you mean there is some randomness involved ?

Reply:
Yes - because the point displacement renderer uses a spatial index the order of points/groups rendered isn't set. I **think** that's the only one which is affected by this.

Reply:
Ok ... I am not very familiar with this one. I'll have to test to understand how it impacts masking.